homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,304 @@
+})
+
+FILTER_THROTTLE_SCHEMA = FILTER_SCHEMA.extend({
+    vol.Required(CONF_FILTER_NAME): FILTER_NAME_THROTTLE,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_FILTERS): vol.All(cv.ensure_list,
+                                        [vol.Any(FILTER_OUTLIER_SCHEMA,
+                                                 FILTER_LOWPASS_SCHEMA,
+                                                 FILTER_THROTTLE_SCHEMA)])
+})
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
+    """Set up the template sensors."""
+    name = config.get(CONF_NAME)
+    entity_id = config.get(CONF_ENTITY_ID)
+    filters = []
+
+    for _filter in config[CONF_FILTERS]:
+        window_size = _filter.get(CONF_FILTER_WINDOW_SIZE)
+        precision = _filter.get(CONF_FILTER_PRECISION)
+
+        if _filter[CONF_FILTER_NAME] == FILTER_NAME_OUTLIER:
+            radius = _filter.get(CONF_FILTER_RADIUS)
+            filters.append(OutlierFilter(window_size=window_size,
+                                         precision=precision,
+                                         entity=entity_id,
+                                         radius=radius))
+        elif _filter[CONF_FILTER_NAME] == FILTER_NAME_LOWPASS:
+            time_constant = _filter.get(CONF_FILTER_TIME_CONSTANT)
+            filters.append(LowPassFilter(window_size=window_size,
+                                         precision=precision,
+                                         entity=entity_id,
+                                         time_constant=time_constant))
+        elif _filter[CONF_FILTER_NAME] == FILTER_NAME_THROTTLE:
+            filters.append(ThrottleFilter(window_size=window_size,
+                                          precision=precision,
+                                          entity=entity_id))
+
+    async_add_devices([SensorFilter(name, entity_id, filters)])
+
+
+class SensorFilter(Entity):
+    """Representation of a Filter Sensor."""
+
+    def __init__(self, name, entity_id, filters):
+        """Initialize the sensor."""
+        self._name = name
+        self._entity = entity_id
+        self._unit_of_measurement = None
+        self._state = None
+        self._filters = filters
+        self._icon = ICON
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def filter_sensor_state_listener(entity, old_state, new_state):
+            """Handle device state changes."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+            self._icon = new_state.attributes.get(ATTR_ICON, self._icon)
+
+            if new_state.state in [STATE_UNKNOWN, STATE_UNAVAILABLE]:
+                return
+
+            self._state = new_state.state
```

Comment:
I find this a very weird assignment as it is never used. I would prefer a local variable that is assigned to `self._state` at the end. That way if something unexpected blows up, you are not left with a temp variable as state.

Reply:
its not temp, its just the state of the previous filter

Reply:
`new_state.state` is the state that triggered the state change. It has not been filtered yet. Since we have at least 1 filter, this assignment will never last when we update Home Assistant.

Reply:
I would expect only to write values to `self._state` that you are intending to publish to Home Assistant. Class instance variables should not be used for temporary variables inside the scope.